kernel/src/fs/epoll/entry.rs

```diff
@@ -0,0 +1,381 @@
+    /// event masks and flags.
+    fn set_enabled(&self, _guard: &MutexGuard<Inner>) {
+        self.is_enabled.store(true, Ordering::Relaxed)
+    }
+
+    /// Marks the epoll entry as not enabled.
+    ///
+    /// This method must be called while holding the lock of the event masks and flags. This is the
+    /// only way to ensure that the "is enabled" state describes the correct combination of the
+    /// event masks and flags.
+    fn reset_enabled(&self, _guard: &MutexGuard<Inner>) {
+        self.is_enabled.store(false, Ordering::Relaxed)
+    }
+
+    /// Gets an instance of `Weak` that refers to the epoll entry.
+    fn weak_entry(&self) -> &Weak<Entry> {
+        &self.weak_entry
+    }
+}
+
+impl events::Observer<IoEvents> for Observer {
+    fn on_events(&self, _events: &IoEvents) {
+        self.ready_set.push(self);
+    }
+}
+
+/// A set of ready epoll entries.
+pub(super) struct ReadySet {
+    // Entries that are probably ready (having events happened).
+    entries: SpinLock<VecDeque<Weak<Entry>>, LocalIrqDisabled>,
+    // A guard to ensure that ready entries can be popped by one thread at a time.
+    pop_guard: Mutex<PopGuard>,
+    // A pollee for the ready set (i.e., for `EpollFile` itself).
+    pollee: Pollee,
+}
+
+struct PopGuard;
+
+impl ReadySet {
+    pub(super) fn new() -> Self {
+        Self {
+            entries: SpinLock::new(VecDeque::new()),
+            pop_guard: Mutex::new(PopGuard),
+            pollee: Pollee::new(IoEvents::empty()),
+        }
+    }
+
+    pub(super) fn push(&self, observer: &Observer) {
+        // Note that we cannot take the `Inner` lock because we are in the callback of the event
+        // observer. Doing so will cause dead locks due to inconsistent locking orders.
+        //
+        // We don't need to take the lock because
+        // - We always call `file.poll()` immediately after calling `self.set_enabled()` and
+        //   `file.register_observer()`, so all events are caught either here or by the immediate
+        //   poll; in other words, we don't lose any events.
+        // - Catching spurious events here is always fine because we always check them later before
+        //   returning events to the user (in `Entry::poll`).
+        if !observer.is_enabled() {
+            return;
+        }
+
+        let mut entries = self.entries.lock();
+
+        if !observer.is_ready() {
+            observer.set_ready(&entries);
+            entries.push_back(observer.weak_entry().clone())
+        }
+
+        // Even if the entry is already set to ready,
+        // there might be new events that we are interested in.
+        // Wake the poller anyway.
+        self.pollee.add_events(IoEvents::IN);
+    }
+
+    pub(super) fn drain(&self) -> ReadySetDrain {
+        ReadySetDrain {
+            ready_set: self,
+            _pop_guard: self.pop_guard.lock(),
+            limit: None,
+        }
```

Comment:
How about renaming this method as `hold_pop`? Indeed, this method does not tries to drain the `ready_set`, but holds the `pop_guard`.

Reply:
I think it behaves like [`Vec::drain`](https://doc.rust-lang.org/std/vec/struct.Vec.html#method.drain)? There seems to be only a small difference: With `Vec`, if we drop `vec::Drain`, all items to be drained are also dropped, but here, if we drop `ReadySetDrain`, the items to be drained are not affected.

`hold_pop` seems very confusing to me. Maybe `lock_pop` is better, but I'm not sure how to name the type `ReadySetDrain` then.

Reply:
> With Vec, if we drop vec::Drain, all items to be drained are also dropped, but here, if we drop ReadySetDrain, the items to be drained are not affected.

Yes, this is why I feel using `Drain` is a bit strange, because it has a different meaning.

> hold_pop seems very confusing to me. Maybe `lock_pop` is better,

Yes, lock_pop looks fine to me

> but I'm not sure how to name the type ReadySetDrain then.

How about `ReadySetPopIter`?

Reply:
These names seem reasonable to me. I've renamed `drain` to `pop_lock` and `ReadySetDrain` to `ReadySetPopIter`.